drain3/template_miner.py

```diff
@@ -170,24 +172,46 @@ def match(self, log_message: str, full_search_strategy="never") -> LogCluster:
         matched_cluster = self.drain.match(masked_content, full_search_strategy)
         return matched_cluster
 
-    def get_parameter_list(self, log_template: str, content: str):
+    def get_parameter_list(self, log_template: str, content: str, valid_mask_generator: Callable[[str], str] = None):
+        if not valid_mask_generator:
+            def valid_mask_generator(old_mask):
+                return old_mask + random.choice(old_mask)
+        masking_parameters = {}
         escaped_prefix = re.escape(self.config.mask_prefix)
         escaped_suffix = re.escape(self.config.mask_suffix)
-        template_regex = re.sub(escaped_prefix + r".+?" + escaped_suffix, self.drain.param_str, log_template)
-        if self.drain.param_str not in template_regex:
-            return []
-        template_regex = re.escape(template_regex)
-        template_regex = re.sub(r'\\ +', r'\\s+', template_regex)
-        template_regex = "^" + template_regex.replace(escaped_prefix + r"\*" + escaped_suffix, "(.*?)") + "$"
+        content_masked = content
+        for match in re.finditer(escaped_prefix + r"(.+?)" + escaped_suffix, content_masked):
+            # Mark masks already present in the message as containing None as parameter.
+            # This way they cannot be confused with any real parameters.
+            masking_parameters[match.group(1)] = None
+        for instruction in self.config.masking_instructions:
+            def mask_parameter(match):
+                id = instruction.mask_with
+                while id in masking_parameters.keys():
+                    id = valid_mask_generator(id)
+                masking_parameters[id] = match.group(0)
+                return self.config.mask_prefix + id + self.config.mask_suffix
+            content_masked = instruction.regex.sub(mask_parameter, content_masked)
```

Comment:
Parameters are masked with a unique mask, that can later be looked up to get the actual parameter.

To make sure other masking-patterns do not interfere with the added mask, an optional function `valid_mask_generator` can be provided, that should generate a new valid mask (and is provided with the mask of an unsuccessful attempt to create an unique mask).
The default for `valid_mask_generator` should work in most cases, as it just adds another random character from the existing mask.